database/skylink.go

```diff
@@ -137,9 +144,38 @@ func (db *DB) AddServerForSkylinks(ctx context.Context, skylinks []string, serve
 	} else {
 		update = bson.M{"$addToSet": bson.M{"servers": server}}
 	}
-	opts := options.Update().SetUpsert(true)
-	_, err := db.staticDB.Collection(collSkylinks).UpdateMany(ctx, filter, update, opts)
-	return err
+
+	var sls []string
+	for len(skylinks) > 0 {
+		if len(skylinks) > maxNumSkylinksToProcess {
+			sls, skylinks = skylinks[:maxNumSkylinksToProcess], skylinks[maxNumSkylinksToProcess:]
+		} else {
+			sls, skylinks = skylinks[:], []string{}
+		}
+
+		// Make sure the skylink records exist.
+		type simpleSkylinkRec struct {
+			Skylink string `bson:"skylink"`
+		}
+		recs := make([]interface{}, len(sls), len(sls))
+		for idx, sl := range sls {
+			recs[idx] = simpleSkylinkRec{Skylink: sl}
+		}
+		insOps := options.InsertMany().SetOrdered(false)
+		_, err := db.staticDB.Collection(collSkylinks).InsertMany(ctx, recs, insOps)
+		if err != nil && !mongo.IsDuplicateKeyError(err) {
+			return err
+		}
+
+		// Update the skylink records.
+		filter := bson.M{"skylink": bson.M{"$in": sls}}
+		_, err = db.staticDB.Collection(collSkylinks).UpdateMany(ctx, filter, update)
```

Comment:
Upserting here led to a record in the DB with a missing skylink field. I can play around with it a bit more.

Reply:
I played with this for about an hour. There's probably some aggregation magic that can be done but that will take me forever to figure out on my own. Attacking this directly with something like 
```
db.getCollection('test').updateMany({s:{"$in":["a","b","s"]}}, {"$set":{n:1}}, {upsert:true})
```
sets the `n` field on existing records to 1 but:
1. if there are records that match the filter then it updates them and ignores the rest
2. if there aren't records that match the filter, it inserts a single record with no `s` field

Neither of these are what we want. I suggest we stay with the current (admittedly inefficient) implementation because it's simple enough for us to verify that it's correct and reliable. We can come back to this and optimise it in the future.

Reply:
I'm less worried about it not being efficient and more worried about race conditions. e.g. UpdateMany could fail because something deletes the links in the meantime. 
Since we are probably not deploying today we could ask PJ on Monday if he got a clue and if not merge it as is for now.

Reply:
I don't think it's possible for this to fail for that reason but I'm happy to ask PJ to look into it.

Reply:
I'll merge this, so we can move forward. If PJ has actionable input here I'll add a F/U.

Reply:
you can't upsert multiple documents, the approach would be to insert and update many without upsert imo